src/framework/MOM_diag_buffers.F90

```diff
@@ -0,0 +1,513 @@
+  write(stdout,*) '==== MOM_diag_buffers: diag_buffers_unit_tests_2d ==='
+  fail = fail .or. new_buffer_2d()
+  fail = fail .or. grow_buffer_2d()
+  fail = fail .or. fill_value_2d()
+  fail = fail .or. store_buffer_2d()
+  fail = fail .or. reuse_buffer_2d()
+
+  contains
+
+  !> Ensure properties of a newly initialized buffer
+  function new_buffer_2d() result(local_fail)
+    type(diag_buffer_2d) :: buffer
+    logical :: local_fail !< True if any of the unit tests fail
+    local_fail = .false.
+    local_fail = local_fail .or. allocated(buffer%buffer)
+    if (verbose) write(stdout,*) "new_buffer_2d: ", local_fail
+    local_fail = local_fail .or. allocated(buffer%ids)
+    if (verbose) write(stdout,*) "new_buffer_2d: ", local_fail
+    local_fail = local_fail .or. buffer%length /= 0
+    if (verbose) write(stdout,*) "new_buffer_2d: ", local_fail
+  end function new_buffer_2d
+
+  !> Test the growing of a buffer
+  function grow_buffer_2d() result(local_fail)
+    type(diag_buffer_2d) :: buffer
+    logical :: local_fail !< True if any of the unit tests fail
+    integer, parameter :: is=1, ie=2, js=3, je=6
+    integer :: i
+
+    local_fail = .false.
+
+    call buffer%set_horizontal_extents(is=is, ie=ie, js=js, je=je)
+    ! Grow the buffer 3 times
+    do i=1,3
+      call buffer%grow()
+      local_fail = local_fail .or. (buffer%length /= i)
+      local_fail = local_fail .or. (lbound(buffer%buffer(i)%field, 1) /= is)
+      local_fail = local_fail .or. (ubound(buffer%buffer(i)%field, 1) /= ie)
+      local_fail = local_fail .or. (lbound(buffer%buffer(i)%field, 2) /= js)
+      local_fail = local_fail .or. (ubound(buffer%buffer(i)%field, 2) /= je)
+    enddo
+    if (verbose) write(stdout,*) "grow_buffer_2d: ", local_fail
+  end function grow_buffer_2d
+
+  !> Test that growing new buffer fills the array with a set fill value
+  function fill_value_2d() result(local_fail)
+    type(diag_buffer_2d) :: buffer
+    logical :: local_fail !< True if any of the unit tests fail
+    integer, parameter :: is=1, ie=2, js=3, je=6
+    real, parameter :: fill_value = -123.456
+    integer :: i
+
+
+    local_fail = .false.
+
+    call buffer%set_horizontal_extents(is=is, ie=ie, js=js, je=je)
+    call buffer%set_fill_value(fill_value)
+    ! Grow the buffer 3 times
+    call buffer%grow()
+    if (any(buffer%buffer(1)%field(:,:) /= fill_value)) local_fail = .true.
+    if (verbose) write(stdout,*) "fill_value_2d: ", local_fail
+  end function fill_value_2d
+
+  !> Test storing a buffer based on a unique id
+  function store_buffer_2d() result(local_fail)
+    type(diag_buffer_2d) :: buffer
+    logical :: local_fail !< True if any of the unit tests fail
+
+    integer, parameter :: is=1, ie=2, js=3, je=6, nlen=3
+    integer :: i, slot
+    real, allocatable, dimension(:,:,:) :: test_2d
+
+    local_fail = .false.
+
+    allocate(test_2d(nlen, is:ie, js:je))
+    call random_number(test_2d)
+    buffer = diag_buffer_2d(is=is, ie=ie, js=js, je=je)
```

Comment:
This is causing an error in the Nvfortran compiler:
```
NVFORTRAN-F-0155-No default initialization in structure constructor- member buffer (../../../../.././/src/MOM6/src/framework/MOM_diag_buffers.F90: 326)
```
This is happening in nvhpc 23.9.

Not sure exactly what is causing the error, but the structure constructor may be unsupported.  Can you explicitly assign these `is`/`ie`/`js`/`je` values?